### ISSUE TITLE
[Question] Memory usage is not reduced by deleting points

### ISSUE THREAD

**Issue by joeltorby** [qdrant/qdrant#3273]
Hi,

I'm new to Qdrant and have created a collection with python inspired by [Qdrant tutorial](https://qdrant.tech/documentation/guides/optimize/#prefer-low-memory-footprint-with-high-speed-search):

```
client = QdrantClient(...)

client .recreate_collection(
              collection_name="test",
              vectors_config=VectorParams(size=1536, distance=Distance.COSINE),
              optimizers_config=models.OptimizersConfigDiff(memmap_threshold=20000),
              quantization_config=models.ScalarQuantization(
                  scalar=models.ScalarQuantizationConfig(
                      type=models.ScalarType.INT8,
                      always_ram=True,
                  ),
          )
  )
```

This works as expected when I upload points to the collection, and the memory usage goes up.

But when I delete points from the collection, the memory usage doesn't drop as I would expect. 

I'm guessing this is something about the `always_ram=True`, Memmap-memory or cache. I wonder if there is a function or something I can run after I have deleted points to reset the ram-memory/Memmap-memory/chache to make the collection only keep the non-deleted data in memory. 

Thanks in advance!

**Comment by generall** [comment]
Hi @joeltorby, measuring memory usage is more tricky than it looks. Most likely what you see is a disk cache. Usually it is normal and won't cause OOM errors if configured properly

**Comment by timvisee** [comment]
To add to this: deleting a point doesn't always mean that the point data is immediately deleted from disk. That's because it's usually more efficient to do it in bulk when an optimizer picks it up.

This section describes some basics of it: https://qdrant.tech/documentation/concepts/collections/#approximate-point-and-vector-counts

So I'm saying that: this may be expected behavior, and don't forget that measuring memory usage may not be very accurate as @generall mentiond above.

**Comment by joeltorby** [comment]
Thanks for your answers @generall & @timvisee, that was somewhat reassuring to hear. 

Here's a brief update on the situation: the memory usage of my Qdrant node is shown in the graph below. As you can see, I added a significant number of points on December 22nd, which I then quickly deleted. Based on your previous response, which mentioned that data "doesn't .. immediately deleted" and the documentation referring to "temporarily duplicate points", I was expecting the memory usage to stabilize around 1.2 GB. However, that doesn't seem to be happening.

Could this be an expected behavior, or is it more likely that there is a misconfiguration somewhere?

![Screenshot 2024-01-04 121611](https://github.com/qdrant/qdrant/assets/39265201/56ba1a95-54b4-41e7-8718-377e93062461)

**Comment by timvisee** [comment]
> Could this be an expected behavior, or is it more likely that there is a misconfiguration somewhere?

Yes, I think this can be expected. It depends on the current state of your collection(s) and on how they're used since then.

Let me try to clarify what 'temporary' and 'not immediately' means: each collection has a bunch of segments, these segments contain the data that is in your collection. When you insert a point, it is added to one of these segments. Qdrant also has an optimizer, which tries to _optimize_ these segments when certain thresholds are reached.

Usually, when you delete a point it isn't deleted right away but instead it's 'marked for deletion'. When the optimizer picks segments up, these points get removed. Note that because of this, removal only happens when the optimization threshold is reached and the optimizer picks such segment up.

Again, it is expected behavior for points to remain for some (or even a long) time. Generally speaking, the more updates you have on your collection, the more likely it is for the optimizer to do work.

I mentioned 'thresholds': these are the optimizer thresholds you can configure for a collection, such as the indexing threshold. Generally speaking I recommend to leave these at their defaults.

Looking at the above graph the reason for this utilization pattern could also be on-disk (memmap) vectors, payload data or indexes. Are you using on-disk by any chance? Also, since memory usage floats at around 2.5G, what is the total memory you have available on this machine?
